### PR TITLE
chore: separate tsconfig.build.json for building, enable more strict …

### PR DESCRIPTION
**Pull request by rpiszczatowski**
…compilation rules in tsconfig.json

### CODE REVIEW

File: packages/cache-service/tsconfig.build.json [hunk 1]
```diff
@@ -1,4 +1,23 @@
...
+      "path": "../sdk/tsconfig.build.json"
+    },
+    {
+      "path": "../streamr-proxy/tsconfig.build.json"
+    },
+    {
+      "path": "../utils/tsconfig.build.json"
+    }
+  ]
+}
```
Comment (cehali) [inline packages/cache-service/tsconfig.build.json:23]: Please add new line

File: packages/cache-service/tsconfig.json [hunk 2]
```diff
@@ -1,39 +1,45 @@
...
-      "path": "../streamr-proxy"
+      "path": "../streamr-proxy/tsconfig.build.json"
     },
     {
-      "path": "../utils"
+      "path": "../utils/tsconfig.build.json"
     }
   ]
-}
+}
```
Comment (cehali) [inline packages/cache-service/tsconfig.json:45]: Please add new line

File: packages/evm-connector/tsconfig.json [hunk 3]
```diff
@@ -1,34 +1,42 @@
...
-      "path": "../protocol"
+      "path": "../protocol/tsconfig.build.json"
     },
     {
-      "path": "../sdk"
+      "path": "../sdk/tsconfig.build.json"
     }
   ]
-}
+}
```
Comment (cehali) [inline packages/evm-connector/tsconfig.json:42]: Please add new line

File: packages/gelato-relayer/web3-functions/IterationArgsProviderInterface.ts [hunk 4]
```diff
@@ -4,7 +4,7 @@ import { providers } from "ethers";
 export type IterationArgs<Args> = {
   shouldUpdatePrices: boolean;
   args?: Args;
-  message?: string;
+  message?: string | undefined;
```
Comment (cehali) [inline packages/gelato-relayer/web3-functions/IterationArgsProviderInterface.ts:7]: Is `message?: string` and `message?: string | undefined` the same thing? `message?: string` resolves to `message: string | undefined`

File: packages/cache-service/src/streamr-listener/streamr-listener.service.ts [hunk 5]
```diff
@@ -78,10 +78,10 @@ export class StreamrListenerService {
     this.logger.log(`Stream exists. Connecting to: ${streamId}`);
     const subscription = await this.streamrClient.subscribe(
       streamId,
-      async (message: Uint8Array) => {
+      async (message: unknown) => {
```
Comment (cehali) [inline packages/cache-service/src/streamr-listener/streamr-listener.service.ts:81]: In that case `unknown` is unecessary

File: packages/integration/src/framework/cache-layer-manager.ts [hunk 6]
```diff
@@ -17,12 +17,12 @@ import {
...
-  mongoDbProcess?: ChildProcess;
-  directCacheServiceProcess?: ChildProcess;
-  publicCacheServiceProcess?: ChildProcess;
+  mongoDbProcess?: ChildProcess | undefined;
+  directCacheServiceProcess?: ChildProcess | undefined;
+  publicCacheServiceProcess?: ChildProcess | undefined;
   directCacheServicePort?: number;
   publicCacheServicePort?: number;
-  dotenvPath?: string;
+  dotenvPath?: string | undefined;
```
Comment (cehali) [inline packages/integration/src/framework/cache-layer-manager.ts:25]: Do we need to add `| undefined`? mongoDbProcess?: ChildProcess resolves to mongoDbProcess: ChildProcess | undefined

File: packages/on-chain-relayer/hardhat.config.ts [hunk 7]
```diff
@@ -10,7 +10,7 @@ dotenv.config();
 
 const accounts = process.env.PRIVATE_KEY
   ? [process.env.PRIVATE_KEY]
-  : undefined;
+  : undefined!;
```
Comment (cehali) [inline packages/on-chain-relayer/hardhat.config.ts:13]: Maybe change `undefined!` to `[]`. It is weird to add `!` to `undefined` 😄

File: packages/oracle-node/src/fetchers/SlippageAndLiquidityCommons.ts [hunk 8]
```diff
@@ -20,7 +20,12 @@ export const tryConvertUsdToTokenAmount = (
   decimalsMultiplier: number,
   amountInUsd: number
 ): string | undefined => {
-  const rawPrice = getRawPrice(assetId)?.value;
+  let rawPrice = undefined;
+  try {
+    rawPrice = getRawPrice(assetId)?.value;
+  } catch (e: unknown) {
+    // ignore
+  }
```
Comment (cehali) [inline packages/oracle-node/src/fetchers/SlippageAndLiquidityCommons.ts:28]: What is the purpose of this change? `getRawPrice` is not async function

File: packages/on-chain-relayer/src/types.ts [hunk 9]
```diff
@@ -33,8 +33,8 @@ export interface OnChainRelayerManifest {
 
 export interface RelayerConfig {
   relayerIterationInterval: number;
-  updatePriceInterval?: number;
-  cronExpressions?: string[];
+  updatePriceInterval?: number | undefined;
```
Comment (cehali) [inline packages/on-chain-relayer/src/types.ts:36]: Do we need to add `| undefined`?

File: packages/oracle-node/src/types.ts [hunk 10]
```diff
@@ -84,14 +84,14 @@ export interface PricesObj {
 export interface PricesObjWithMetadata {
   [symbol: string]: {
     value: PriceDataFetchedValue;
-    metadata?: RedstoneTypes.MetadataPerSource;
+    metadata?: RedstoneTypes.MetadataPerSource | undefined;
```
Comment (cehali) [inline packages/oracle-node/src/types.ts:87]: Do we need to add `| undefined`

File: packages/oracle-node/tsconfig.build.json [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      "path": "../oracles-smartweave-contracts/tsconfig.build.json"
+    },
+    {
+      "path": "../streamr-proxy/tsconfig.build.json"
+    },
+    {
+      "path": "../utils/tsconfig.build.json"
+    }
+  ]
+}
```
Comment (cehali) [inline packages/oracle-node/tsconfig.build.json:29]: Please add new line

File: packages/oracle-node/tsconfig.json [hunk 12]
```diff
@@ -24,25 +30,28 @@
...
-      "path": "../streamr-proxy"
+      "path": "../streamr-proxy/tsconfig.build.json"
     },
     {
-      "path": "../utils"
+      "path": "../utils/tsconfig.build.json"
     }
   ]
-}
+}
```
Comment (cehali) [inline packages/oracle-node/tsconfig.json:57]: Please add new line

File: packages/oracles-smartweave-contracts/tsconfig.build.json [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "./tsconfig.json",
+  "exclude": [
+    "node_modules",
+    "test",
+    "dist",
+    "**/*spec.ts"
+  ]
+}
```
Comment (cehali) [inline packages/oracles-smartweave-contracts/tsconfig.build.json:9]: Please add new line

File: packages/oracles-smartweave-contracts/tsconfig.json [hunk 14]
```diff
@@ -1,26 +1,37 @@
...
     "src/contracts/redstone-oracle-registry/initial-state.json"
   ],
-  "exclude": ["node_modules", "tools", "test"]
-}
+  "exclude": [
+    "node_modules",
+    "tools",
+    "test"
+  ]
+}
```
Comment (cehali) [inline packages/oracles-smartweave-contracts/tsconfig.json:37]: Please add new line

File: packages/protocol/tsconfig.build.json [hunk 15]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "./tsconfig.json",
+  "exclude": [
+    "node_modules",
+    "test",
+    "dist",
+    "**/*spec.ts"
+  ]
+}
```
Comment (cehali) [inline packages/protocol/tsconfig.build.json:9]: Please add new line

File: packages/protocol/src/data-point/DataPoint.ts [hunk 16]
```diff
@@ -6,7 +6,7 @@ import { INumericDataPoint } from "./NumericDataPoint";
 export interface IStandardDataPoint {
   dataFeedId: ConvertibleToBytes32;
   value: string; // base64-encoded bytes
-  metadata?: Metadata;
+  metadata?: Metadata | undefined;
```
Comment (cehali) [inline packages/protocol/src/data-point/DataPoint.ts:9]: Do we need to add `| undefined`?

File: packages/protocol/src/data-point/NumericDataPoint.ts [hunk 17]
```diff
@@ -12,9 +12,9 @@ import { DataPoint, Metadata } from "./DataPoint";
 export interface INumericDataPoint {
   dataFeedId: ConvertibleToBytes32;
   value: number;
-  decimals?: number;
+  decimals?: number | undefined;
```
Comment (cehali) [inline packages/protocol/src/data-point/NumericDataPoint.ts:15]: Do we need to add `| undefined`?

File: packages/protocol/tsconfig.json [hunk 18]
```diff
@@ -1,21 +1,31 @@
...
+    "target": "ES2021",
   },
-  "include": ["src/**/*.ts", "test", "test/**/*.json"]
-}
+  "include": [
+    "src/**/*.ts",
+    "test",
+    "test/**/*.json"
+  ]
+}
```
Comment (cehali) [inline packages/protocol/tsconfig.json:31]: Please add new line

File: packages/rpc-providers/src/TransactionDeliveryMan.ts [hunk 19]
```diff
@@ -69,7 +69,7 @@ export type TransactionDeliveryManOpts = {
   /**
    * Should be set to true if chain doesn't support EIP1559
    */
-  isAuctionModel?: boolean;
+  isAuctionModel?: boolean | undefined;
```
Comment (cehali) [inline packages/rpc-providers/src/TransactionDeliveryMan.ts:72]: Do we need to add `| undefined`?

File: packages/rpc-providers/tsconfig.build.json [hunk 20]
```diff
@@ -1,23 +1,13 @@
...
+    "./hardhat.config.ts"
+  ],
   "references": [
     {
-      "path": "../utils"
+      "path": "../utils/tsconfig.build.json"
     }
   ]
-}
+}
```
Comment (cehali) [inline packages/rpc-providers/tsconfig.build.json:13]: Please add new line

File: packages/sdk/src/index.ts [hunk 21]
```diff
@@ -20,9 +20,9 @@ export interface DataPackagesRequestParams {
   dataServiceId: string;
   uniqueSignersCount: number;
   dataFeeds?: string[];
-  urls?: string[];
+  urls?: string[] | undefined;
```
Comment (cehali) [inline packages/sdk/src/index.ts:23]: Do we need to add `| undefined`?

File: packages/sdk/tsconfig.build.json [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+      "path": "../protocol/tsconfig.build.json"
+    },
+    {
+      "path": "../oracles-smartweave-contracts/tsconfig.build.json"
+    },
+    {
+      "path": "../utils/tsconfig.build.json"
+    }
+  ]
+}
```
Comment (cehali) [inline packages/sdk/tsconfig.build.json:20]: Please add new line

File: packages/sdk/tsconfig.json [hunk 23]
```diff
@@ -1,31 +1,41 @@
...
-      "path": "../oracles-smartweave-contracts"
+      "path": "../oracles-smartweave-contracts/tsconfig.build.json"
     },
     {
-      "path": "../utils"
+      "path": "../utils/tsconfig.build.json"
     }
   ]
-}
+}
```
Comment (cehali) [inline packages/sdk/tsconfig.json:41]: Please add new line

File: packages/streamr-proxy/tsconfig.build.json [hunk 24]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "./tsconfig.json",
+  "exclude": [
+    "node_modules",
+    "test",
+    "dist",
+    "**/*spec.ts"
+  ]
+}
```
Comment (cehali) [inline packages/streamr-proxy/tsconfig.build.json:9]: Please add new line

File: packages/streamr-proxy/tsconfig.json [hunk 25]
```diff
@@ -1,21 +1,30 @@
...
-    "incremental": true
+    "target": "ES2021",
   },
-  "include": ["./src", "./test"]
-}
+  "include": [
+    "./src",
+    "./test"
+  ]
+}
```
Comment (cehali) [inline packages/streamr-proxy/tsconfig.json:30]: Please add new line

File: packages/utils/src/types/redstone-metadata.ts [hunk 26]
```diff
@@ -12,7 +12,7 @@ export interface MetadataPerSource {
   slippage?: SlippageData[];
 
   /** Defined only for sources which supports fetching liquidity. */
-  liquidity?: string;
+  liquidity?: string | undefined;
```
Comment (cehali) [inline packages/utils/src/types/redstone-metadata.ts:15]: Do we need to add `| undefined`?

File: packages/utils/tsconfig.build.json [hunk 27]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "./tsconfig.json",
+  "exclude": [
+    "node_modules",
+    "test",
+    "dist",
+    "**/*spec.ts"
+  ]
+}
```
Comment (cehali) [inline packages/utils/tsconfig.build.json:9]: Please add new line

File: packages/utils/tsconfig.json [hunk 28]
```diff
@@ -1,23 +1,32 @@
...
+    "target": "ES2021",
   },
   "include": [
     "./src/**/*.ts",
     "./test/**/*.ts",
     "package.json",
     "src/**/*.json"
   ]
-}
+}
```
Comment (cehali) [inline packages/utils/tsconfig.json:32]: Please add new line

**Review by cehali — DISMISSED** [review]
Minor things in comments

File: packages/cache-service/tsconfig.build.json (hunk 1 above)
Comment (rpiszczatowski) [reply packages/cache-service/tsconfig.build.json:23]: To do that we need to change our prettier. This is what it is doing while you try to format the file. I am not sure if we want to do it.

File: packages/cache-service/src/streamr-listener/streamr-listener.service.ts (hunk 5 above)
Comment (rpiszczatowski) [reply packages/cache-service/src/streamr-listener/streamr-listener.service.ts:81]: It is necessary because the subscribe method expects us to pass a callback that accepts `unknown` as a type. Remember that this is streamrClient that will call this callback. They can pass whatever they want according to `subscribe` signature (because unknown is literally anything). Even though the method shouldn't be called with a type other than `Uint8Array` this is not what the interface promises.

File: packages/integration/src/framework/cache-layer-manager.ts (hunk 6 above)
Comment (rpiszczatowski) [reply packages/integration/src/framework/cache-layer-manager.ts:25]: Yes we do. Another option is to rewrite usages not to explicitly assign undefined to those properties. Look at my explanation above to the other similar question.

File: packages/on-chain-relayer/hardhat.config.ts (hunk 7 above)
Comment (rpiszczatowski) [reply packages/on-chain-relayer/hardhat.config.ts:13]: [] is a correct option. But I wanted to leave the generated code as it was. And I like `undefined!` ;)
I'll of course apply the suggestion.

File: packages/oracle-node/src/fetchers/SlippageAndLiquidityCommons.ts (hunk 8 above)
Comment (rpiszczatowski) [reply packages/oracle-node/src/fetchers/SlippageAndLiquidityCommons.ts:28]: You are right. It throws when the last price is too old. I stumbled upon the situation when my fetcher stopped working because of that. Method `tryXXX` shouldn't throw because it reports lack of value with `undefined` return value.

File: packages/cache-service/src/bundlr/bundlr.service.ts [hunk 29]
```diff
@@ -8,7 +8,7 @@ const REDSTONE_TYPE_TAG_VALUE = "redstone-oracles";
 @Injectable()
 export class BundlrService {
   private readonly logger = new Logger(BundlrService.name);
-  private bundlrClient: Bundlr;
+  private bundlrClient: Bundlr | undefined;
```
[the remark this answers is not in the document]
Comment (rpiszczatowski) [reply packages/cache-service/src/bundlr/bundlr.service.ts:11]: As above. This is not the same.

File: packages/cache-service/tsconfig.build.json (hunk 1 above)
Comment (rpiszczatowski) [reply packages/cache-service/tsconfig.build.json:23]: My bad. I was using a different formatted. Will fix it.

File: packages/gelato-relayer/web3-functions/IterationArgsProviderInterface.ts (hunk 4 above)
Comment (koonopek) [reply packages/gelato-relayer/web3-functions/IterationArgsProviderInterface.ts:7]: Is it sensible from run time perspective or this is just TS gymnastic that we have to do, because of how TS works?

Comment (rpiszczatowski) [reply packages/gelato-relayer/web3-functions/IterationArgsProviderInterface.ts:7]: Good question! Those two situations (property existing and equal to undefined and property not existing) are distinguishable at runtime. `obj.hasOwnProperty("property")` returns true or false depending on that. So we may feel the difference while doing
```
for (const p in args) {
  console.log(`${p} -> ${args[p]}`);
}
```
or
```
if ('message' in args) {
  console.log("message in args");
} else {
  console.log("message not in args");
}
```

If, on the other hand, we simply do `args.message` there is no difference in what the resulting value is, we will get `undefined` in both cases.

My perspective is that once you wrap your head around this distinction, it is even cleaner, meaning that now, with a higher level of errors, I feel I understand better what the compiler says to me :). Maybe this is because I had some experience with vanilla JS some time ago.

**Review by koonopek — APPROVED** [review]
GJ remember to disable this exactProperty rule ;)
